src/main/java/uk/gov/hmcts/reform/em/orchestrator/service/caseupdater/CcdBundleCloningService.java

```diff
@@ -0,0 +1,85 @@
+package uk.gov.hmcts.reform.em.orchestrator.service.caseupdater;
+
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.hmcts.reform.em.orchestrator.service.ccdcallbackhandler.CcdCallbackDto;
+import uk.gov.hmcts.reform.em.orchestrator.service.dto.CcdBundleDTO;
+import uk.gov.hmcts.reform.em.orchestrator.service.dto.CcdValue;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@Transactional
+public class CcdBundleCloningService implements CcdCaseUpdater {
+
+    private final ObjectMapper objectMapper;
+    private final JavaType type;
+
+    public CcdBundleCloningService(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+        type = objectMapper.getTypeFactory().constructParametricType(CcdValue.class, CcdBundleDTO.class);
+    }
+
+    @Override
+    public boolean handles(CcdCallbackDto ccdCallbackDto) {
+        return false;
+    }
+
+    @Override
+    public JsonNode updateCase(CcdCallbackDto ccdCallbackDto) {
```

Comment:
There's a few things in this method I would change:

- the inner loop should be extracted to a private method
- the updatedBundlesList is only used inside the if statement so it should be defined there
- the maybeBundles.removeAll followed by .addAll is a sign you are replacing (mapping!) the value, which is nice because...
- when dealing with Options the preferred approach is map the result rather than for if statements, for example:

```
List<JsonNode> newBundles = ccdCallbackDto
    .findCaseProperty(ArrayNode.class)
    .map(CcdBundleCloningService::getClonedBundleList)

ccdCallbackDto.setBundles(newBundles);

return ccdCallbackDto.getCaseData();
```

There's a couple of things to consider around error handling - you can either throw the IOException in your inner map method and have to wrap it with an `unchecked` function, but I would just return the original list if you hit an error.

Reply:
actually the example I gave can be improved, will chat after stand up